### ISSUE TITLE
Premake Project for Premake

### ISSUE THREAD

**Issue by NyxFaust** [TheCherno/Hazel#241]
**By Adding the following to the end of your premake5.lua main file you can then re-run Premake and you will gain a Premake project which will give you easy editing access to all your Premake files within the project. You can rebuild the Premake project and it will run Premake for you and update your projects on the fly.**

project "Premake"
	kind "Utility"

	files
	{
		"**.lua"
	}
	
	buildmessage 'Re-generating Project Files!'

	postbuildcommands
	{
		'%{prj.location}/vendor/premake/premake5 vs2017'
	}

**Hope this helps some of you who dislike having to edit from the directory or running the batch file.**

**Comment by LovelySanta** [comment]
Not sure how good this would be since they are all called premake5? If they are easy to distinguish, i think this could indeed be a neat feature, otherwise not.

**Comment by NyxFaust** [comment]
each premake is in the library folder its contained within making them easy to distinguish.

**Comment by NyxFaust** [comment]
https://gyazo.com/b26694acb740797db4812ba650e3a8b0

**Comment by LovelySanta** [comment]
I approve, this would be nice to have. However, same comment as in https://github.com/TheCherno/Hazel/issues/242#issuecomment-629277977: wait till cherno made the game into a seperate project, then we can implement these two issues together.

**Comment by LovelySanta** [comment]
Also note that we also want to implement #52 at the same time.

**Comment by Gaztin** [comment]
I know this is not a PR, but something to note is that we should use [`_ACTION`](https://github.com/premake/premake-core/wiki/_ACTION) for anyone not running exactly `vs2017`.

**Comment by NyxFaust** [comment]
@Gaztin 
How would you use the _ACTION global from within Visual Studio? The initial run of the project sure but I believe it would be simpler for the project file to simply be updated to 2019 if that is what is being used. sometimes KISS does come into play. But if you have a suggestion for full automation of the process?

**Comment by Gaztin** [comment]
It would be resolved the same way `%{prj.location}` will be resolved. At initial run of the premake script.
All you'd have to do is replace
```lua
'%{prj.location}/vendor/premake/premake5 vs2017'
```
with
```lua
'%{prj.location}/vendor/premake/premake5 %{_ACTION}'
```

**Comment by Gaztin** [comment]
And on the topic of programming principles, remember cross-platform compatibility. Visual Studio is not a thing on Linux.

**Comment by NyxFaust** [comment]
@Gaztin 
When I run your changes via the project build within VS it fails it works from the command line however.

I understand VS is not a thing within Linux however I tend to compile Linux binaries quite often through VS for my server code. I'm not a Linux or Apple specialist, In this case I'm simply sharing a concept which is why i did not put it as a Pushed bit of code but rather under Issues.

**Comment by Gaztin** [comment]
This is more of a proof-of-concept deal, I take it.

Well, the concept of not having to run the premake configuration manually is convenitn, but I feel like it's a slippery slope. For instance, premake allows for custom options that can be passed via the command line. Suddenly we are faced with an issue where we cannot enforce people to use custom options for, let's say, setting their PhysX SDK location.

That said, I don't really think it's any significant inconvenience to keep a terminal open and very occasionally press upwards and return.

**Comment by NyxFaust** [comment]
@Gaztin 
I get you and when making a high use library or following standardization in processing your right but were talking a project that frankly will change and morph over time with users from a wide range of backgrounds who will inevitably break that concept by making things there own.. In that case being frank here 90%+ plus will be under windows and them having the option to build for that platform and recreate the projects in a convenient manner is not opening a console finding a folder and typing a command every time they close the console window or restart having a IDE specific way of redoing projects via a single key press is more convenient in my mind. But that is just me we all do things the way that works best for us and our own productivity.

**Comment by LovelySanta** [comment]
Why not have both worlds? instead of naming it `GenProjects` since it's windows specific, why not have a `GenVS2017Projects` and a `GenVS2019Projects` at that point? Or prompt the user to press key 1 or 2 depending on the version (and maybe press enter/wait a second for default option) It's not like there are so many options to choose from...

**Comment by NyxFaust** [comment]
@LovelySanta 
That could work or you could simply have individual batch files, and or shell files for the posix users out there.

**Comment by Gaztin** [comment]
Yeah I mean in short retrospect, I think you're right about this being a convenience update. What I'm still unsure about is why we can't just use `_ACTION` and suddenly we have a solution that is not only cross platform, but cross vs-versions.

**Comment by LovelySanta** [comment]
I just remember, in #154 I managed to [extract the version of VS that is installed](https://github.com/LovelySanta/Hazel/blob/dfd5bb7926a30f0f27e8ee6df7465809bbc507d5/scripts/Win-BuildPremake.bat#L24-L71) on the system, so we could use that in the bat file to build the correct VS files?

**Comment by NyxFaust** [comment]
@LovelySanta 
Yeah I don't see why not!

**Comment by Gaztin** [comment]
I think that is unnecessarily complicated. Unless you want to re-generate vs2019-files when you've originally generated them as vs2017. Which will lead to bad behavior if you happen to have vs2019 installed, but are mainly using vs2017.

**Comment by LovelySanta** [comment]
I'm trying to implement this, in addition to #301, but I'm having a few difficulties.

Creating the project on it's own 'works' if I add it to main `premake.lua` file. However I do not want the `Premake.vcxproj` files to generate in the root directory (same location as the `Hazel.sln` location). I found 2 solutions, initially I did set the location of the premake project to vendor/premake. That fixed the location initially, the file structure is shown below. After a while in #301 I moved on to just making the premake project a seperate file as I did with the other files, such that it just sits inside `vendor/premake`, such that I didn't have to tinker with the location at all.
![image](https://user-images.githubusercontent.com/26593477/89712516-9b776580-d991-11ea-9bd1-0943c7b3e9f4.png)

However when I try to run the "rebuild" command inside VS, it doesn't work as intended. I get the message
```
1>------ Rebuild All started: Project: Premake, Configuration: Debug x64 ------
1>EXEC : error : ...ns/Documents/dev/Hazel_Santa/vendor/premake/premake5.lua:1: no workspace in scope
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(148,5): error MSB3073: The command "C:\Users\Laurens\Documents\dev\Hazel_Santa\vendor/premake/bin/premake5 vs2019
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\Microsoft.CppCommon.targets(148,5): error MSB3073: :VCEnd" exited with code 1.
1>Done building project "Premake.vcxproj" -- FAILED.
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
```

So it does find the `premake.exe` file, but it doesn't execute it from the root directory (workspace). So we should cd into the workspace directory before running the premake command itself. This is the same thing the script file does. This is the point that I'm a bit stuck on.

My premake project currently looks like
```lua
project "Premake"
	kind "Utility"

	targetdir ("%{wks.location}/bin/" .. outputdir .. "/%{prj.name}")
	objdir ("%{wks.location}/bin-int/" .. outputdir .. "/%{prj.name}")

	files
	{
		"%{wks.location}/**premake5.lua"
	}

	buildmessage 'Regenerating Project Files!'

	postbuildcommands
	{
		"%{wks.location}vendor/premake/bin/premake5 %{_ACTION}"
	}
```

And the project structure looks like
![image](https://user-images.githubusercontent.com/26593477/89712470-53584300-d991-11ea-89f1-ef02de8d87e6.png)

I did not add this to #301 yes as it's in a not working state as of now. Any help would be welcome :)

**Comment by SentientCoffee** [comment]
Just gonna paraphrase the discussion on Discord:
@LovelySanta's problem is that when he runs the post-build command for the Premake project which runs the premake5.exe file, the premake5.lua file that is beside the executable is being used first instead of the one at the Hazel repo's root directory, resulting in a project that is outside a workspace (which errors out).

The suggestion was to use the `--file` command when running premake5.exe:
```lua
postbuildcommands
{
    "%{wks.location}vendor/premake/bin/premake5 %{_ACTION} --file=\"path/to/file.lua\""
}
```

which will allow Lovely_Santa to run Premake with the premake5.lua located in the root directory first which will create the workspace and associate all the projects with one solution file.

**Comment by LovelySanta** [comment]
As @SentientCoffee suggested, in https://github.com/TheCherno/Hazel/pull/301/commits/47f2bdb3d9a96e365a9149a5ac6001e0baa225d2 the postbuild was slightly altered to include the --file parameter, such that the main premake file will be used, which is located in the workspace location. Also slightly changed the path to the executable, as this can easily be relative to the project file.

I also solved the build message as I changed it a postbuild message as it's a postbuild command:
https://github.com/TheCherno/Hazel/blob/47f2bdb3d9a96e365a9149a5ac6001e0baa225d2/vendor/premake/premake5.lua#L12-L16

Rebuilding inside VS now outputs the correct result:
![image](https://user-images.githubusercontent.com/26593477/89715325-3463ac00-d9a5-11ea-8e49-269f2b0b3da5.png)
